[FLINK-11328] [core] Migrate all parameterless serializers to use new serialization compatibility API

tzulitai opened the pull request:
## What is the purpose of the change

This PR migrates all parameterless serializers (in Flink, they happen to all be subclasses of `TypeSerializerSingleton`) to use the new serialization compatibility APIs, so that the serializers themselves are no longer Java-serialized into savepoints.

## Brief change log

- edf6d59: This commit essentially does 2 things: 1) improve usability of the new `SimpleTypeSerializerSnapshot` base class, so that it may be used by serializers that have varying ways for instantiation, and 2) let all subclasses of `TypeSerializerSingleton` return their own implementation of a `SimpleTypeSerializerSnapshot`, while removing the base implementation of `snapshotConfiguration` in `TypeSerializerSingleton`.
- 898f6c5: Some CEP serializers for NFA data structures use nested serializers, so they should actually use the `CompositeTypeSerializerSnapshot` snapshot class. This commit addresses that.
- 74adb84 to b4910d4: Some cleanup in tests.
- 79c85ca to 18187c0: Commits that add migration tests for the touched serializers.

## Verifying this change

All new subclasses of `TypeSerializerSnapshotMigrationTestBase` should pass.

tzulitai commented:
@dawidwys 
Regarding the changes to the CEP serializers, could you take a look at 898f6c54035206a97cfbb04bdad4f7568264dab2?

dawidwys commented:
Hi @tzulitai I had a look just at the commit you linked. I've put some suggestions. Most of it is just nitpicking, the biggest concern I have is maybe we should still return a fixed size for `EventIdSerializer`

tzulitai commented:
@dawidwys I've addressed your comments in fixup commit 6c7a3fe.
Could you take a look and let me know if there are other issues, or a +1 if it looks good?

dawidwys commented:
Thanks @tzulitai for the update. It looks good for me now. +1

tzulitai commented:
Thanks! Will merge after Travis confirms green ..

tzulitai commented:
Merging ...
